File: arm/api/forms/v_lk_curator/v_lk_curator.js
//
//
//
window.sovaActions = window.sovaActions || {};
window.sovaActions.v_lk_curator = {
	init2: doc => {
		if (doc.fieldValues.GRID)
			doc.util.getJson(doc, `form=v_lk_curator&event=${doc.getField('event')}&cmd=getSgr&group=${doc.fieldValues.GRID}`)
				.then( sgr => { 
					doc.changeDropList('leftList', sgr, 0);
					setTimeout(() => doc.forceUpdate(), 100);
				})
				.catch( () => {});
	},
	// *** *** ***

	hide: {
		viewbar: doc => !doc.checked,
	},
	// *** *** ***

	cmd: {
		newPay: (doc, par, ctrl, shiftKey) => {
			let [sgrId, sstId, pref] = par.split('|');
			
			let view = doc.getControl('mainList');
			view.selectedDoc = sstId;
			view.forceUpdate();
						
			let page = {
				addUrl: `&sgrId=${sgrId}&sstId=${sstId}&profile=${pref}`,			
				rsMode: 'new',
				newForm: 'Payment',
				dbAlias: 'nv_Payment',
				title: 'Новый платеж',
			};
			doc.previewNew(page);
		},
		
		selOne: doc => {
			doc.checked = false;
			for (let k in doc.register) {
				if (k.startsWith('SELONE_') && doc.getField(k)) {
					doc.checked = true;
					break;
				}
			}
			doc.forceUpdate();
		},
		setField: (doc, prm) => {
			let text;
			let [fi, id, value] = prm.split('|');

			let view = doc.getControl('mainList');
			view.selectedDoc = id;
			view.forceUpdate();

			switch(fi) {
				case 'was_s':
					text = value ? 'Установить "Пропустил"' : 'Установить "Был на занятии"';
					break;
				case 'esse_s':
					text = value ? 'Сбросить "Эссе"' : 'Установить "Эссе"';
					break;
				case 'consultant_s':
					text = value ? 'Сбросить "Консультант"' : 'Установить "Консультант"';
					break;
				case 'allow_s':
					text = value ? 'Сбросить "Допуск"' : 'Установить "Допуск"';
					break;
			}
			doc.msg.box('выбранному студенту ?', `Установка полей|${text}\n`)
				.then( () => {
					let body = JSON.stringify([`form=v_lk_curator&cmd=setField&field=${fi}`, `${id}=${value ? '' : '1'}`]);
					doc.util.getJson(doc, body, true)
						.then( res => {
							if (res !== 'OK')
								console.error(`http-status: ${res}`)
							doc.loadView('mainList', true, id);
						})
						.catch( err => console.error(err) );
				}) // end msg.box
				.catch(() => {});
		},
		cmdSet: (doc, fi) => {
			let text;
			let value = '1';
			if (fi === 'allow_s')
				text = 'Установить "допуск"';
			else if (fi === 'allow_r') {
				text = 'Сбросить "допуск"';
				fi = 'allow_s';
				value = '';
			}
			else if (fi === 'was_s')
				text = 'Установить "Был на занятии"';
			else if (fi === 'was_r') {
				text = 'Установить "Пропустил"';
				fi = 'was_s';
				value = '';
			}
			else
				text = 'Установить "зачет"';

			doc.msg.box('выбранным студентам ?', `Установка полей|${text}\n`)
				.then( () => {
					let buf = '';
					let id;
					for (let k in doc.register) {
						if (k.startsWith('SELONE_') && doc.getField(k)) {
							id = doc.util.partition(k, '_')[1];
							if (buf)
								buf += '¤';
							buf += `${id}=${value}`;
						}
					}
					if (buf) {
						let body = JSON.stringify([`form=v_lk_curator&cmd=setField&field=${fi}`, buf]);
                        doc.util.getJson(doc, body, true)
							.then( res => {
								if (res !== 'OK')
									console.error(`http-status: ${res}`)
								doc.loadView('mainList', true);
							})
							.catch( err => console.error(err) );
					}
				}) // end msg.box
				.catch(() => {});
		},
		cmdPref:(doc, pk, ctrlKey, shiftKey) => {
			let [pkPref, pkSst] = pk.partition('|');
			let view = doc.getControl('mainList');
			view.rowClick(pkSst);
			let page = {form: 'Profile', dbAlias: 'nv_Profile', rsMode: 'edit', unid: pkPref};
			page.title = 'Редактирование профайла студента';
			doc.previewNew(page, ctrlKey, shiftKey);
		},

		cmdEdit:(doc, pk, ctrlKey, shiftKey) => {
			let view = doc.getControl('mainList');
			view.rowClick(pk);
			let page = doc.util.urlKeys(view.props.previewUrl);
			page.title = 'Редактирование сессии студента';
			page.unid = pk;
			page.rsMode = 'edit';
			doc.previewNew(page, ctrlKey, shiftKey);
		},
		
		openSession: (doc, value, ctrlKey, shiftKey) => {
			let ls = (value + '|').split('|');
			let page = {
				frameStyle: {'width': 1000},
				title: doc.getField('group') || 'Сессия группы',
				unid: ls[1],
				dbAlias: 'nv_SessionGr',
				rsMode: 'edit',
			};
			doc.previewNew(page, ctrlKey, shiftKey);
		},

		// *** *** ***

	},
	recalc: {
		EVENT: (doc, val) => {
			if (doc.fieldValues.GRID)
				doc.util.getJson(doc, `form=v_lk_curator&event=${val}&cmd=getSgr&group=${doc.fieldValues.GRID}`)
					.then( sgr => { 
						doc.changeDropList('leftList', sgr, 0);
						setTimeout(() => doc.forceUpdate(), 100);
					})
					.catch( () => {});
		},

		SELECTALL: (doc, val) => {
			let checked = false;
			for (let k in doc.register) {
				if (k.startsWith('SELONE_')) {
					doc.setField(k, val);
					checked = val;
				}
			}
			doc.checked = checked;
			doc.forceUpdate();
		},

		STATUS: doc => {
			if (document.getElementById('chb_vm'))
				document.getElementById('chb_vm').checked = doc.checked = false;
			doc.loadView('mainList', true);
		},


		LEFTLIST: (doc, selectedLeft) => {
			if (document.getElementById('chb_vm'))
				document.getElementById('chb_vm').checked = doc.checked = false;
			let mainList = doc.getControl('mainList');
			mainList.mainDocs = [];
			for (let k in doc.register) {
				if (k.startsWith('SELONE_'))
					delete doc.register[k];
			} 
			mainList.loadView(true); // true -> refresh
		},

		UPLIST: (doc, selectedLeft) => {
			if (document.getElementById('chb_vm'))
				document.getElementById('chb_vm').checked = doc.checked = false;
			
			let grId = doc.getField('upList').partition('|')[1];
			doc.util.getJson(doc, `cmd=well&clues=sessionsGr_GrId_band|${grId}`)
				.then( sgr => doc.changeDropList('leftList', sgr, 0))
				.catch( () => {});				
		},
	}
};

window.sovaActions.v_lk_curator2 = window.sovaActions.v_lk_curator;

File: arm/api/forms/v_lk_curator/v_lk_curator.py
# -*- coding: utf-8 -*-
'''
Created on 2020.

@author: aon
'''
from arm.tools.first import err
from arm.tools.DC import well, swell, DC, toWell
from arm.tools.loadWell import loadWell
from arm.api.forms.formTools import style, _div, _btnD, _field, _btnEdit, _btnPref, gridStyle
from arm.api.forms.classPage import Page
from arm.tools.dbToolkit.DJ import docFromDB
from arm.api.forms.toolbars import toolbar
from arm.api.forms.sstButtons import sstButtons

from django.http import HttpResponse

import json

# *** *** ***


class v_lk_curator(Page):
    '''
    список сессий студентов для конкретоной sGr
    Установка/сброс "Д-З-Р-V" для выделенных студней
    Открывается из лк куратора по кнопке слева "номер группы"(под * * *)
    Слева список сессий, справа студенты
    '''
    dbAlias = 'nv_SessionSt'
    _VIEW_ = 1

    def __init__(self, request):
        self.form = getattr(self, '__module__', '').rpartition('.')[2]
        self.jsCssUrl = [f'/api/jsv?forms/v_lk_curator/v_lk_curator.js']
        self.leftWidth = 150

        super().__init__(request)

    def getData(self, dcUK):
        if not (dcUK._staff or 'куратор' in dcUK._role):
            return '""'
        if dcUK.cmd == 'setField':
            try:
                flag = None
                toWell(1, 'busy')
                i = 0
                for ch in dcUK.buf.split('¤'):
                    if ch:
                        pk, _, val = ch.partition('=')
                        dc = DC(unid=pk, dbAlias='nv_SessionSt', fullName=dcUK.fullName, _superUser=dcUK._superUser)
                        i += 1
                        if docFromDB(dc) and dc.form != 'SessionGr':  # в списке могут быть сиссии группы(commonSessGr)
                            if dc.doc[dcUK['field']] != val:
                                dc.doc[dcUK['field']] = val
                                dc.save()
                                flag = True
                toWell(0, 'busy')
                if flag:
                    sgr = well('sessionGr_Id', dc.doc.sessionGr)
                    loadWell('SessionSt', sgr.nvgroup_id)
                return HttpResponse('OK')

            except Exception as ex:
                toWell(0, 'busy')
                loadWell('SessionSt')
                err(f'getData (cmd={dcUK.cmd}):{ex}', cat=self.form)
                return HttpResponse(f'getData for {self.form}(cmd={dcUK.cmd}): {ex}', None, 200)

        # ***

        elif dcUK.cmd == 'getSelected':
            data = self.getView(dcUK)

        # CH GROUP
        elif dcUK.cmd == 'changeStatus':
            data = swell('sessionsGr_GrId_band', dcUK.groupdId) or []
            if dcUK.status != '1':
                data = [s for s in data if s.endswith('|active')]

        elif dcUK.cmd == 'getSgr':
                if dcUK.event:
                    ls = well('sessionsGr_GrId', dcUK.group)
                    data = [f'{dc.title}|{dc.id}' for dc in ls if dc.nvEvent == dcUK.event]
                else:  # v_lk_curator2
                    data = swell('sessionsGr_GrId_band', dcUK.group) or []
        else:
            data = [f'invalid cmd: {dcUK.cmd}']

        return json.dumps(data, ensure_ascii=False)

    # *** *** ***
    setButtons = [
                _field('selectAll', 'chb', [''], chbView='nv', edit=1, className='chbVN'),
                _div(**style(width=10), name='viewbar'),
                _btnD('Допуск+', 'cmdSet', 'allow_s', name='viewbar'),
                _div(**style(width=10), name='viewbar'),
                _btnD('Доп.-', 'cmdSet', 'allow_r', name='viewbar'),
                _div(**style(width=10), name='viewbar'),
                _btnD('Был', 'cmdSet', 'was_s', name='viewbar'),
                _div(**style(width=10), name='viewbar'),
                _btnD('Не был', 'cmdSet', 'was_r', name='viewbar'),
                # _div(**style(width=10), name='viewbar'),
                # _btnD('Зачёт+', 'cmdSet', 'test_s', name='viewbar'),
    ]

    def page(self, request):
        if self._userAgent == 'mobile':
            event = _field(f'event', 'lbsd', list(['Все|'] + swell('events')),
                alias=1, edit=1, xValue='Все',
                title='выберите событие',
                name=f'event',
                **style(width=230, margin='auto'),
            )
        else:
            event = _field(f'event', 'band', swell('shortEvents'), recalcText=1,
                className='radioBand',
                **style(margin='auto'),
                title='выберите событие',
                name=f'event')

        status = _div(
            **style(margin='auto',),
            children=[_field('status', 'band', ['актив', 'все'],)
        ])

        self.upField = _div(children=[
            _div(className='toolbar', children=[toolbar.close_]),
            event,
        ])
        self.viewbar = self.makeViewbar(leftBtn=[status], rightBtn=self.setButtons)
        self.leftList = _field('leftList', 'band', [], className='list3str')
        return self.shamrock(addUrl='&status={status}')

    # *** *** ***

    def getView(self, dcUK):
        mainDocs = []

        if dcUK.selected:
            sgrId = dcUK.selected.partition('|')[0]
        else:  # v_lk_curator2
            sgrId = dcUK.sgrId

        sessArr = well('sessionSt_sgrId', sgrId)
        if not sessArr:
            return {'mainDocs': [('1', 'Сессии для студентов не созданы (нет даты начала)'), ]}

        sgr = well('sessionGr_Id',sgrId)
        gridStr = ''
        for sst in sessArr:
            if dcUK.status == '0' and sst.status != 'active':
                continue

            prof = well('profiles', sst.pref)
            if prof.status != 'active':
                continue

            if sst.other_group:
                color = '#888'
                s = f"(подмена в {well('groups_groupId', sst.other_group).title})"
            elif sst.owner:
                color = '#f55'
                s = f"(подмена из {well('groups_groupId', sst.owner).title})"
            else:
                color = '#000'
                s = ''
                
            title = _div(f"{prof.full_name}\n{sgr.d2} {s}",
                s2=1, br=1, **style(letterSpacing=1, paddingLeft=2, color=color))

            pk = sst.id
            chb = _field(f'selOne_{pk}', 'chb', [''], cmd='selOne', chbView='nv', className='checkboxFV')
            btnE = _btnEdit('cmdEdit', pk)
            btnP = _btnPref('cmdPref', f'{sst.pref}|{pk}')

            div1 = _div(children=[chb, title])

            sstBtn = sstButtons(sst, sgr, curator=True)
            gridStr = gridStr or f'{"32px "*(len(sstBtn)+2)}'
            div2 = _div(**gridStyle(gridStr),
                children=[*sstBtn, btnE, btnP])

            row = _div(className='lk_row', children=[
                _div(className='lk_left', children=[div1]),
                _div(className='lk_right', children=[div2])
            ])

            mainDocs.append([pk, row, prof and prof.full_name])

        mainDocs = [[x[0], x[1]] for x in sorted(mainDocs, key=lambda x: x[2])]
        return {'mainDocs': mainDocs, 'refsDocs': None}

    def queryOpen(self, r):
        dcUK = r.dcUK
        self.title = dcUK.title or 'ЛК куратора'
        dcUK.doc.group = dcUK.group or dcUK.title
        dcUK.doc.grId = dcUK.unid

# *** *** ***

